File: 1631-path-with-minimum-effort/1631-path-with-minimum-effort.java
class Solution {
    public int minimumEffortPath(int[][] heights) {
        
        int rows = heights.length;
        int cols = heights[0].length;
        
        int[][] dirs = new int[][] {{1,0},{-1,0}, {0,1},{0,-1}};

        boolean[][] visited = new boolean[rows][cols];

        //row, col, cost
        PriorityQueue<int[]> minHeap = new PriorityQueue<>((a,b) -> Integer.compare(a[2],b[2]));
        minHeap.offer(new int[] {0, 0, 0});
        visited[0][0] = true;

        while (minHeap.size() > 0)
        {
            int[] entry = minHeap.poll();
            int row = entry[0], col = entry[1], cost = entry[2];

            if (row == rows - 1 && col == cols - 1) return cost;

            visited[row][col] = true;

            for (int[] dir : dirs)
            {
                int newRow = row + dir[0];
                int newCol = col + dir[1];

                if (0 <= newRow && newRow < rows
                    && 0 <= newCol && newCol < cols
                    && !visited[newRow][newCol])
                {
                    int diff = Math.abs(heights[row][col] - heights[newRow][newCol]);
                    int newCost = Math.max(cost, diff);
                    minHeap.offer(new int[] {newRow, newCol, newCost});
                }
            }
        }

        return -1;
    }
}

File: 1631-path-with-minimum-effort/1631-path-with-minimum-effort.py
class Solution:
    def minimumEffortPath(self, heights: List[List[int]]) -> int:
        rows, cols = len(heights), len(heights[0])
        minHeap = [(0, 0, 0)] # (abs diff, r, c)
        direct = [[0, 1], [1, 0], [-1, 0], [0, -1]]
        visit = set()

        while minHeap:
            diff, r, c = heapq.heappop(minHeap)
            if r == rows - 1 and c == cols - 1:
                return diff

            if (r, c) in visit:
                continue
        
            visit.add((r, c))

            for dr, dc in direct:
                nr, nc = r + dr, c + dc
                if 0 <= nr < rows and 0 <= nc < cols and (nr, nc) not in visit:
                    new_diff = max(diff, abs(heights[r][c] - heights[nr][nc]))
                    heapq.heappush(minHeap, (new_diff, nr, nc))
        